### PR TITLE
Remove staticroute finalizer

### PR DESCRIPTION
**Pull request by TaoZou1**
The finalizer may block the deletion since the NSX resource may take long time to delete. 
Remove the finalizer so the deleting operation will return immediately

Test Done:
1. create a staticroute, check if it has no finalizer
2. delete the staticroute, check log if the staticroute has been deleted from nsx side

### CODE REVIEW

File: pkg/controllers/staticroute/staticroute_controller.go [hunk 1]
```diff
@@ -72,60 +71,86 @@ func deleteSuccess(r *StaticRouteReconciler, _ context.Context, o *v1alpha1.Stat
...
-				return ResultRequeue, err
-			}
-			log.V(1).Info("removed finalizer", "staticroute", req.NamespacedName)
-			deleteSuccess(r, ctx, obj)
-		} else {
-			// only print a message because it's not a normal case
-			log.Info("finalizers cannot be recognized", "staticroute", req.NamespacedName)
-		}
+	if !obj.ObjectMeta.DeletionTimestamp.IsZero() {
+		return r.deleteStaticRoute(ctx, req, obj)
```
Comment (timdengyun) [inline pkg/controllers/staticroute/staticroute_controller.go:140]: For this case, you can get CR when deleting, so you can keep logic
if err := r.Service.DeleteStaticRoute(obj); err != nil {
				log.Error(err, "delete failed, would retry exponentially", "staticroute", req.NamespacedName)
				deleteFail(r, ctx, obj, &err)
				return ResultRequeue, err
			}
			deleteSuccess(r, ctx, obj)

File: pkg/controllers/staticroute/staticroute_controller.go [hunk 2]
```diff
@@ -72,28 +69,61 @@ func deleteSuccess(r *StaticRouteReconciler, _ context.Context, o *v1alpha1.Stat
...
+		log.Info("deleting StaticRoute", "StaticRouteUID", uid, "nsxStaticRouteId", *item.Id)
+		path := strings.Split(*item.Path, "/")
+		if err := r.Service.DeleteStaticRouteByPath(path[2], path[4], path[6], *item.Id); err != nil {
+			log.Error(err, "failed to delete StaticRoute", "StaticRouteUID", uid, "nsxStaticRouteId", *item.Id)
+			return err
+		}
+		log.Info("successfully deleted StaticRoute", "StaticRouteUID", uid, "nsxStaticRouteId", *item.Id)
+	}
+	return nil
+}
```
Comment (wenqiq) [inline pkg/controllers/staticroute/staticroute_controller.go:108]: Should add a space line here between functions?

File: pkg/controllers/staticroute/staticroute_controller_test.go [hunk 3]
```diff
@@ -366,3 +338,111 @@ func TestStaticRouteReconciler_Start(t *testing.T) {
 	err := r.Start(mgr)
 	assert.NotEqual(t, err, nil)
 }
+func TestStaticRouteReconciler_listStaticRouteCRIDs(t *testing.T) {
```
Comment (wenqiq) [inline pkg/controllers/staticroute/staticroute_controller_test.go:341]: Better add a space line.

File: pkg/nsx/services/staticroute/staticroute.go [hunk 4]
```diff
@@ -147,7 +148,18 @@ func (service *StaticRouteService) DeleteStaticRoute(obj *v1alpha1.StaticRoute)
...
+	staticroutes := service.StaticRouteStore.GetByIndex(common.TagScopeNamespace, ns)
+	for _, staticroute := range staticroutes {
+		sr := staticroute.(*model.StaticRoutes)
+		tagname := nsxutil.FindTag(sr.Tags, common.TagScopeStaticRouteCRName)
+		if tagname == name {
+			result = append(result, staticroute.(*model.StaticRoutes))
+		}
+	}
+	return result
+}
```
Comment (wenqiq) [inline pkg/nsx/services/staticroute/staticroute.go:162]: ditto.

File: pkg/nsx/services/staticroute/staticroute.go [hunk 5]
```diff
@@ -147,7 +148,18 @@ func (service *StaticRouteService) DeleteStaticRoute(obj *v1alpha1.StaticRoute)
 	}
 	return service.DeleteStaticRouteByPath(vpcResourceInfo.OrgID, vpcResourceInfo.ProjectID, vpcResourceInfo.VPCID, id)
 }
-
+func (service *StaticRouteService) ListStaticRouteByName(ns, name string) []*model.StaticRoutes {
```
Comment (wenqiq) [inline pkg/nsx/services/staticroute/staticroute.go:151]: ditto

File: pkg/controllers/staticroute/staticroute_controller.go [hunk 6]
```diff
@@ -72,28 +69,61 @@ func deleteSuccess(r *StaticRouteReconciler, _ context.Context, o *v1alpha1.Stat
...
-	}
-
-	if obj.ObjectMeta.DeletionTimestamp.IsZero() {
-		metrics.CounterInc(r.Service.NSXConfig, metrics.ControllerUpdateTotal, common.MetricResTypeStaticRoute)
-		if !controllerutil.ContainsFinalizer(obj, commonservice.StaticRouteFinalizerName) {
-			controllerutil.AddFinalizer(obj, commonservice.StaticRouteFinalizerName)
-			if err := r.Client.Update(ctx, obj); err != nil {
-				log.Error(err, "add finalizer", "staticroute", req.NamespacedName)
-				updateFail(r, ctx, obj, &err)
+		if client.IgnoreNotFound(err) == nil {
```
Comment (wenqiq) [inline pkg/controllers/staticroute/staticroute_controller.go:115]: Use `apierrors.IsNotFound(err)` for a better understanding. It has been changed in other removing finalizer PRs.

**Review by timdengyun — DISMISSED** [review]
LGTM

**Review by wenqiq — APPROVED** [review]

**Review by timdengyun — APPROVED** [review]
LGTM

**Review by wenqiq — APPROVED** [review]
LGTM

**Review by timdengyun — DISMISSED** [review]
LGTM

**Review by wenqiq — APPROVED** [review]

**Review by timdengyun — APPROVED** [review]
LGTM

**Review by wenqiq — APPROVED** [review]
LGTM

**Pull request merged**